pde_superresolution/WENO.py

```diff
@@ -0,0 +1,111 @@
+import numpy as np
+from .equations import Grid, RandomForcing, ConservativeBurgersEquation, staggered_first_derivative
+
+class WENO(object):
+
+  def __init__(self, k: int, dx: float):
+    if k not in (2, 3):
+      raise NotImplementedError('WENO is only implemented for k=2,3')
+    self.k = k
+    self.dx = dx
+    self.epsilon = 1e-6
+
+    # precompute coefficients
+    self.c = {r: np.array([WENO._c(r, j, self.k) for j in range(self.k)]) for r in range(-1, self.k)}
+    self.d = {r: np.array([WENO._d(r, j, self.k) for j in range(self.k)])/self.dx for r in range(-1, self.k)}
+
+    if k == 2:
+      self.gamma = np.array([2, 1]) / 3
+    if k == 3:
+      self.gamma = np.array([3, 6, 1]) / 10
+
+  '''
+  Reconstruction coefficients for the function value. Identifies with Table 2.1 in:
+  Essentially non-oscillatory and weighted essentially non-oscillatory schemes for hyperbolic conservation laws.
+  Shu, Chi-Wang
+  Advanced numerical approximation of nonlinear hyperbolic equations. Springer, Berlin, Heidelberg, 1998.
+  https://www3.nd.edu/~zxu2/acms60790S13/Shu-WENO-notes.pdf
+  '''
+  @staticmethod
+  def _c(r: int, j: int, k: int):
+    return np.sum([
+      np.sum([
+        np.prod([
+          r - q + 1 for q in range(k + 1) if q != m and q != l
+        ])
+        for l in range(k + 1) if m != l]
+      ) / np.prod([(m - l) for l in range(k + 1) if l != m])
+      for m in range(j + 1, k + 1)
+    ])
+
+  '''
+  Reconstruction coefficients for the function first derivative. Based on _c but homebrewed with no reference.
+  '''
+  @staticmethod
+  def _d(r: int, j: int, k: int):
+    return np.sum([
+      np.sum([
+        np.sum([
+          np.prod([
+            r - q + 1 for q in range(k + 1) if q != m and q != l and q != alpha
+          ]) for alpha in range(k + 1) if alpha != m and alpha != l])
+        for l in range(k + 1) if m != l]
+      ) / np.prod([(m - l) for l in range(k + 1) if l != m])
+      for m in range(j + 1, k + 1)
+    ])
+
+  '''
+  Calculates the smoothness estimators returns an array of size (k+1, length(y)).
+  The (i,r) element corresponds to smoothness estimates of cell j with offset r.
+  '''
+
+  def smoothness(self, y):
+    if self.k == 2:
+      return np.array([
+        (np.roll(y, 1) - y) ** 2,
```

Comment:
Are you sure `roll()` is doing the right thing here? I think you need to specify the `axis`. To quote numpy:

axis : int or tuple of ints, optional
Axis or axes along which elements are shifted. By default, the array is flattened before shifting, after which the original shape is restored.
https://docs.scipy.org/doc/numpy-1.15.1/reference/generated/numpy.roll.html